### PR TITLE
[Perf] Use fused ops npu_top_k_top_p

### PR DESCRIPTION
**Pull request by Pr0Wh1teGivee**
### What this PR does / why we need it?
Use fused ops torch_npu.npu_top_k_top_p(logits, p, k) when p and k are not None, otherwise fallback to the original one. The replacement will take place automatically when `VLLM_ASCEND_ENABLE_TOPK_OPTIMIZE=1` .

This patch are using `npu_top_k_top_p` which required torch_npu>=2.5.1.post1.dev20250619

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Tested by DeepSeek R1

### CODE REVIEW

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py [hunk 1]
```diff
@@ -51,6 +52,9 @@ def _apply_top_k_top_p(
     p: torch.Tensor,
     k: torch.Tensor,
 ) -> torch.Tensor:
+    if p is not None and k is not None:
+        return torch_npu.npu_top_k_top_p(logits, p, k)
```
Comment (Yikun) [inline vllm_ascend/patch/worker/patch_common/patch_sampler.py:57]: Which torch_npu version supported this call?

Comment (Pr0Wh1teGivee) [reply vllm_ascend/patch/worker/patch_common/patch_sampler.py:57]: since rc2 b050

**Comment by Yikun** [comment]
1. Please do a rebase because we update the torch version morning, better to do a e2e test.
2. Please add a ut to make sure `npu_top_k_top_p` is called and get expected results in : https://github.com/vllm-project/vllm-ascend/tree/main/tests/ut

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py (hunk 1 above)
Comment (Yikun) [reply vllm_ascend/patch/worker/patch_common/patch_sampler.py:57]: Please update commits msg with publich torch_npu version.

such as: https://mirrors.huaweicloud.com/ascend/repos/pypi/torch-npu/

Comment (Yikun) [reply vllm_ascend/patch/worker/patch_common/patch_sampler.py:57]: This API introduced by [torch_npu-2.5.1.post1.dev20250619](https://mirrors.huaweicloud.com/ascend/repos/pypi/torch-npu/torch_npu-2.5.1.post1.dev20250619-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl)

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py [hunk 2]
```diff
@@ -48,9 +49,12 @@ def apply_min_p(
 
 def _apply_top_k_top_p(
     logits: torch.Tensor,
-    p: torch.Tensor,
     k: torch.Tensor,
+    p: torch.Tensor,
```
Comment (Yikun) [inline vllm_ascend/patch/worker/patch_common/patch_sampler.py:53]: Hmm....So, the order of _apply_top_k_top_p was wrong.

I suggested to keep same order with upstream https://github.com/vllm-project/vllm/blob/9a3b88328f7e434cac35b90ee463de6689f9a833/vllm/model_executor/layers/sampler.py#L398

Please change L98

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py (hunk 1 above)
Comment (Pr0Wh1teGivee) [reply vllm_ascend/patch/worker/patch_common/patch_sampler.py:57]: fixed

**Comment by Pr0Wh1teGivee** [comment]
> 1. Please do a rebase because we update the torch version morning, better to do a e2e test.
> 2. Please add a ut to make sure `npu_top_k_top_p` is called and get expected results in : https://github.com/vllm-project/vllm-ascend/tree/main/tests/ut

fixed

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py (hunk 2 above)
Comment (Pr0Wh1teGivee) [reply vllm_ascend/patch/worker/patch_common/patch_sampler.py:53]: fixed

File: tests/ut/worker/patch_common/test_patch_sampler.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        p = torch.tensor([0.9])
+        generators = {0: torch.Generator()}
+        generators[0].manual_seed(42)
+
+        sampler.forward_native(logits, generators, k, p)
+
+        mock_npu_op.assert_called_once_with(logits, p, k)
+
+
+if __name__ == "__main__":
```
Comment (wangxiyuan) [inline tests/ut/worker/patch_common/test_patch_sampler.py]: no need for this

Comment (Pr0Wh1teGivee) [reply tests/ut/worker/patch_common/test_patch_sampler.py]: fixed

File: tests/ut/worker/patch_common/test_patch_sampler.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+from vllm.v1.sample.ops import topk_topp_sampler
+
+
+class TestTopKTopPSamplerOptimize(unittest.TestCase):
+
+    @mock.patch.dict(os.environ, {"VLLM_ASCEND_ENABLE_TOPK_OPTIMIZE": "1"})
+    @mock.patch("torch_npu.npu_top_k_top_p")
+    def test_npu_topk_topp_called_when_optimized(self, mock_npu_op):
+        import vllm_ascend.patch.worker.patch_common.patch_sampler
+        importlib.reload(vllm_ascend.patch.worker.patch_common.patch_sampler)
```
Comment (Yikun) [inline tests/ut/worker/patch_common/test_patch_sampler.py:16]: Please refer this [1] to patch test and make TestTopKTopPSamplerOptimize based on `TestBase`

[1] https://github.com/vllm-project/vllm-ascend/pull/1386/files#diff-eae86bf6e7a9a6ef5d079fa80ca12e946ecff4e587e5b66d3761f2cc7f6bb9c5R4

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py [hunk 5]
```diff
@@ -48,9 +49,13 @@ def apply_min_p(
 
 def _apply_top_k_top_p(
```
Comment (Yikun) [inline vllm_ascend/patch/worker/patch_common/patch_sampler.py:50]: I trend to rename this `_apply_top_k_top_p` to `apply_top_k_top_p` to avoid confused to keep same with:
https://github.com/vllm-project/vllm/blob/c53fec1fcb27aca9475e55c2d1e74c532f5f0364/vllm/v1/sample/ops/topk_topp_sampler.py#L165

**Review by Yikun — APPROVED** [review]
LGTM, please address comments in a new PR.

**Comment by wangxiyuan** [comment]
`apply_min_p` has been removed from vllm main https://github.com/vllm-project/vllm/commit/48fb076cbc651f655aae8ffdff6930907309b2f7   we'll cleanup this patch code once vllm 0.9.2 is comming. Please update to the newest code if you still want this feature.
